pkg/execution/state/redis_state/lua/includes/gcra.lua

```diff
@@ -1,103 +1,128 @@
+				result["remaining"] = math.floor(next / emission)
+			end
+			result["reset_after"] = ttl
+
+			result["limited"] = true
+
+			return result
+		end
+	end
 
-	return {true, used_burst}
-end
+	ttl = tat - now_ms
+	if origQuantity > 0 then
+		-- update state to new_tat
+		ttl = new_tat - now_ms
+		local expiry = string.format("%d", math.max(ttl / 1000, 1))
+		redis.call("SET", key, new_tat, "EX", expiry)
+	end
+	result["ttl"] = ttl
 
-local function gcraUpdate(key, now_ms, period_ms, limit, burst, capacity)
-  -- calculate emission interval (tau) - time between each token
-  local emission = period_ms / math.max(limit, 1)
-
-  -- retrieve theoretical arrival time
-  local tat = redis.call("GET", key)
-  if not tat then
-    tat = now_ms
-  else
-    tat = tonumber(tat)
-  end
-
-  -- calculate next theoretical arrival time
-  local new_tat
-  if now_ms > tat then
-    new_tat = now_ms + (math.max(capacity, 1) * emission)
-  else
-    new_tat = tat + (math.max(capacity, 1) * emission)
-  end
-
-  if capacity > 0 then
-    local expiry = string.format("%d", period_ms / 1000)
-    if expiry == "0" then
-      expiry = "1"
-    end
-    redis.call("SET", key, new_tat, "EX", expiry)
-  end
-end
+	local next = dvt - ttl
+	if next > -emission then
+		local remaining = math.floor(next / emission)
+		result["remaining"] = remaining
+	end
+	result["reset_after"] = ttl
+	result["next"] = next
 
-local function gcraCapacity(key, now_ms, period_ms, limit, burst)
-  -- calculate emission interval (tau) - time between each token
-  local emission = period_ms / math.max(limit, 1)
+	return result
+end
 
-  -- calculate total capacity in time units
-  local total_capacity_time = emission * (limit + burst)
+-- performs gcra rate limiting for a given key.
+--
+-- Returns true on success, false if the key has been rate limited.
+local function gcra(key, now_ms, period_ms, limit, burst)
+	-- NOTE: we need to admit more than a single item every emission interval, as the queue
+	-- does not follow a uniform arrival rate and we would throttle the majority of queue items,
+	-- leading to significantly lower queue throughput
+	local maxBurst = limit + burst - 1
 
-  -- retrieve theoretical arrival time
-  local tat = redis.call("GET", key)
-  if not tat then
-    tat = now_ms
-  else
-    tat = tonumber(tat)
-  end
+	local res = applyGCRA(key, now_ms, period_ms, limit, maxBurst, 1)
 
-  -- remaining capacity in time units
-  local time_capacity_remain = now_ms + total_capacity_time - tat
+	local used_burst = res["tat"] > now_ms
 
-  -- Convert the remaining time budget back into a number of tokens.
-  local capacity = math.floor(time_capacity_remain / emission)
+	return { res["allowed"], used_burst }
+end
 
-  -- The capacity cannot exceed the defined limit + burst.
-  local final_capacity = math.min(capacity, limit + burst)
+local function gcraUpdate(key, now_ms, period_ms, limit, burst, quantity)
+	local allowRefillingAll = limit + burst - 1
```

Comment:
For key queues, we need to be able to consume all capacity in a single backlogRefill call.

This does impact request smoothing within the period, but ensures we don't reject items even though there is capacity in the period, which is the higher priority.